### ISSUE TITLE
Additional devices

### ISSUE THREAD

**Issue by JanJaapKo** [majki09/domoticz_lg_thinq_plugin#3]
Hi,

I've got an LG washer I'dd like to add using your plugin as basis. I could refactor the whole thing so this plugin supports more than just the AC's, but that would require some work on your end to keep the AC u pand running (I may break some things....)

So do you feel like doing this? If not, I'll just fork

**Comment by majki09** [comment]
Hi Jan!

It's just great that you want to contribute and upgrade this plugin!

Since there is no API V2 documentation available all we need is to get all
fields that API is delivering for the washer. You just need to generate
wideq token file, debug the script and check what fields you will get from
the API. With these we can modify the plugin to have washers support.

You can try by yourself or I can do it for you, but I will need your wideq
token file (I don't have LG washer).

Was your washer detected by the plugin?

I'm pleased to help.


czw., 30 gru 2021, 19:18 użytkownik Jan-Jaap Kostelijk <
***@***.***> napisał:

**Comment by JanJaapKo** [comment]
Hey,

I'm fine with building the washer interface (it won't be the first plugin I'm building). And yes, it was returned as a V2 device.

So my plan is to first explore the API for the washer a bit and then rework the example.py to get the generic part out to do the setup and separate the wrapper functionality for the AC in such a way that I can re-use it for the washer and perhaps later someone else can build the other types (it's not so convenient to do when you're not owning one).

Give me a couple of days and I'll come up with a proposal.

**Comment by bastiknck** [comment]
Hello,

We also have an LG washer and intresed in the washer interface :) @JanJaapKo did you have any luck to get it to work?

Thanks,

Michael

**Comment by JanJaapKo** [comment]
I haven't got any further than playing around a bit, since I've got another plugin I'm fixing first..... Will probably take me about a month or so to get some pilot version to work....... 

So I haven't forgotten about it ;)

**Comment by Artens** [comment]
Hi all, we too are working with dryer & washer combination. I've found this repository by @fison67 that has a v1 and a v2 set of commands for multiple devices, supporting Korean and English languages.

Have a look in the Groovy files for a dryer here [https://github.com/fison67/LG-Connector/blob/master/devicetypes/fison67/lg-dryer.src/lg-dryer.groovy](url)
and the washer here [https://github.com/fison67/LG-Connector/blob/master/devicetypes/fison67/lg-washer-v2.src/lg-washer-v2.groovy](url)
There might be some caveats, as one could doubt whether the url extension on the washer on line 242 is actually the implementation, but who knows :)

**Comment by bastiknck** [comment]
@JanJaapKo any luck or news to get it to work? :)

Thanks

**Comment by TokTias** [comment]
Hi !
I also got a v2 LG Washer/Dryer that is recognized by this plugin.
Is there anything I can do to contribute (I do have some 30+ years experience in C/C++ and I have made some very minor contributions related to MySensors to the domoticz development a few years ago) ?
I assume we "just" need to translate needed parts of the Java-code posted in the link(s) by Artens above into the existing python-plugin-code, right ?

Cheers !

**Comment by JanJaapKo** [comment]
Hi all,

I haven't had much time to spend on this one lately (busy with other plugins and other hobbies as well....) so haven't done anything, basically and do not really foresee to do so anytime soon...... What I saw from first investigation is that the design of the plugin needs to be reworked to accommodate the fact that it needs to handle multiple device types. Now it is simply handling one type (AC). If anyone else wants to have a go, be my guest ;). I'll be happy to help out.

**Comment by StarWing01** [comment]
@majki09 does your plugin currently supports heat exchange boilers like the [WH27S](https://www.lg.com/nl/verwarmingsoplossing/lg-wh27s) ?

**Comment by majki09** [comment]
> @majki09 does your plugin currently supports heat exchange boilers like the [WH27S](https://www.lg.com/nl/verwarmingsoplossing/lg-wh27s) ? I installed and tried it, but got some strange results.

Currently it does not. The thing is that every type of device has to be added separately with parsing it's unique wideq file and API fields. It's easy when I own a device (AC) or somebody will send me his wideq file (like we did with heat pump), otherwise it's hard to implement something new.
Also, there is a way [Artens](https://github.com/majki09/domoticz_lg_thinq_plugin/issues/3#issuecomment-1031868643) mentioned, but I did not have time to try it out yet.

**Comment by devrosx** [comment]
hello just downloaded your plugin (i have LG washer)
example found device:
thinq2 devices: 1
c67409de-e50c-1760-87e8-4cbad734366d: Pračka s předním plněním (WASHER F_V7_Y___W.B_2QEUK / thinq2)

i can send wideq_state.json or if you give me info where to dig i can try.... :)

**Comment by Artens** [comment]
> hello just downloaded your plugin (i have LG washer) example found device: thinq2 devices: 1 <REMOVED ID>: Pračka s předním plněním (WASHER F_V7_Y___W.B_2QEUK / thinq2)

Please edit your message to remove your ID as a safety measure.

**Comment by JanJaapKo** [comment]
Hey @Artens  I'll have a look the coming week into adding the washer.

**Comment by JanJaapKo** [comment]
@Artens , @majki09  the plugin currently takes the users selection for the device type. I think it would be much more powerfull to query the API for the device type, do you agree? We can throw an error for none supported types. As future extension, we can then also control multiple devices from 1 instance of the plugin.

**Comment by majki09** [comment]
> @Artens , @majki09 the plugin currently takes the users selection for the device type. I think it would be much more powerfull to query the API for the device type, do you agree? We can throw an error for none supported types. As future extension, we can then also control multiple devices from 1 instance of the plugin.

Yes! This is good idea in my opinion. I like approaches like this. One instance, many device types. That should look really good. Technically it should be first checked in the API if we have any info about all devices/types available on user's account. Then the script would iterate through all of device types, adding appropriate domoticz instances. And if some device on account is not supported by our plugin it will throw a Warning into the logs. I think this would be the best approach.
I could try to do this when I have some free time, I hope soon.

@JanJaapKo would you mind to create new issue for this (with enhancement label) as it is your idea?
Thanks for this!

**Comment by JanJaapKo** [comment]
Well sure can do that. However, it would (perhaps) overlap with this ticket, since (as far as I can see now) the current implementation seems to be using a different version of the API. So that would require quite some refactoring firstly on the API and secondly on the plugin itself, as we'd need to move to the Domoticz extended framework (where 256 units can be connected per device). 

But I'll focus on getting the washer to work first......

**Comment by Artens** [comment]
Two topics are relevant if I understand everything correctly:
- Currently, a token is copied to set the initial wideq_state.json values and assure correct sequence of token information available in the json file. The wideq_state.json should be kept centrally; I'm unsure whether multiple devices are possibly polled/set/unset with the same tokens? My guess would be yes, as current tokes are not disturbed (I've got two mobile devices and the plugin registered and they are not affected, mobile devices are able to readout without problem across washer and dryer). Would we need to register as developers if a change as suggested is applied? This will make it more cumbersome if others wish to apply the plugin.
- Should the polling for devices happen only once? (i.e. upon registering a device) Thinking about security, it would be beneficial to 'search for known devicetype' instead of polling for a full list. The possibility that someone has >1 washer is in my opinion manageable, ideally one can check custom name to confirm. Or wait until a user logs a defect.

**Comment by JanJaapKo** [comment]
Still working on this topic but progressing slowly. This plugin is quite complex to my opinion, could do with some simplification......

**Comment by vinibali** [comment]
hi there!
i'm also raising my hand to do some work with washers. is there any code related progress so far?
